overlord/servicestate/quota_handlers.go

```diff
@@ -123,54 +124,112 @@ func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
-		// restart services which will happen regardless if we
-		// get rebooted after unlocking the state - if we got
-		// rebooted before unlocking the state, none of the
-		// changes we made to state would be persisted and we
-		// would run through everything above here again, but
-		// the second time around EnsureSnapServices would end
-		// up doing nothing since it is idempotent.  So in the
-		// rare case that snapd gets restarted but is not a
+		// refresh security profiles and restart services which
+		// will happen regardless if we get rebooted after
+		// unlocking the state - if we got rebooted before unlocking
+		// the state, none of the changes we made to state would
+		// be persisted and we would run through everything above
+		// here again, but the second time around EnsureSnapServices
+		// would end up doing nothing since it is idempotent.
+		// So in the rare case that snapd gets restarted but is not a
 		// reboot also record which services do need
 		// restarting. There is a small chance that services
 		// will be restarted again but is preferable to the
 		// quota not applying to them.
-		if err := rememberQuotaStateUpdated(t, appsToRestartBySnap); err != nil {
+		if err := rememberQuotaStateUpdated(t, servicesAffected, refreshProfiles); err != nil {
 			return err
 		}
-
 	}
 
-	if err := restartSnapServices(st, t, appsToRestartBySnap, perfTimings); err != nil {
-		return err
+	if len(servicesAffected) > 0 {
+		ts := state.NewTaskSet()
+		var prevTask *state.Task
+		queueTask := func(task *state.Task) {
+			if prevTask != nil {
+				task.WaitFor(prevTask)
+			}
+			ts.AddTask(task)
+			prevTask = task
+		}
+
+		if refreshProfiles {
+			addRefreshProfileTasks(st, queueTask, servicesAffected)
+		}
+		addRestartServicesTasks(st, queueTask, qc.QuotaName, servicesAffected)
+		snapstate.InjectTasks(t, ts)
 	}
+
 	t.SetStatus(state.DoneStatus)
 	return nil
 }
 
+func addRefreshProfileTasks(st *state.State, queueTask func(task *state.Task), servicesAffected map[*snap.Info][]*snap.AppInfo) *state.TaskSet {
+	ts := state.NewTaskSet()
+	for info := range servicesAffected {
+		setupProfilesTask := st.NewTask("setup-profiles", fmt.Sprintf(i18n.G("Update snap %q (%s) security profiles"), info.SnapName(), info.Revision))
+		setupProfilesTask.Set("snap-setup", &snapstate.SnapSetup{
+			SideInfo: &snap.SideInfo{
+				RealName: info.SnapName(),
+				Revision: info.Revision,
+			},
+		})
+		queueTask(setupProfilesTask)
+	}
+	return ts
+}
+
+func addRestartServicesTasks(st *state.State, queueTask func(task *state.Task), grpName string, servicesAffected map[*snap.Info][]*snap.AppInfo) {
+	getServiceNames := func(services []*snap.AppInfo) []string {
+		var names []string
+		for _, svc := range services {
+			names = append(names, svc.Name)
+		}
+		return names
+	}
+
+	sortedInfos := make([]*snap.Info, 0, len(servicesAffected))
+	for info := range servicesAffected {
+		sortedInfos = append(sortedInfos, info)
+	}
```

Comment:
Ths should also work:
```suggestion
	copy(sortedInfos, servicesAffected)
```

Reply:
This does not work unfortunately, we are creating a list of the map keys, not a copy of the map.

Reply:
Oops, I didn't read carefully enough :-)